### PR TITLE
Order same-factory batches of configurations; generate PIDs with a counter for each factory PID.

### LINKED ISSUE #22058: Unnecessary application expansion on restart
**Describe the bug**  
In limited situations, application binaries will be expanded into the workarea cache after restarting the server even when the application has not changed. This will occur intermittently, and only if a configuration file has been updated before restarting the server. 

**Steps to Reproduce**  
Start a server with an application installed, stop the server, update the timestamp on server.xml, start the server. 

**Expected behavior**  
The server should avoid unnecessary application binary expansion. 

**Diagnostic information:**  
 - OpenLiberty Version: 22.0.0.6 (and likely most previous versions)

**Additional context**

### PR DESCRIPTION
**Pull request by tbitonti**

### CODE REVIEW

**Comment by tbitonti** [comment]
#build

**Comment by tbitonti** [comment]
Test failures:

```
Name
Tests Failures Errors Time-Taken Run-Order

no component
82867 1 4

com.ibm.ws.microprofile.openapi_fat FAT tests
50 1 3
  com.ibm.ws.microprofile.openapi.fat.annotations.AnnotationProcessingTest
  3 1 2 12m07s 1
    testPrivateFieldVisibility_EE10_FEATURES_MicroProfile_60
    1 0 1 6m02s 1 (*1.1)
    testNonPublicFieldVisibility_EE10_FEATURES_MicroProfile_60
    1 1 0 3s 2 (*1.2)
    testRequiredCustomNameField_EE10_FEATURES_MicroProfile_60
    1 0 1 6m02s 3 (*1.3)
  junit.framework.TestSuite
  1 0 1 0s 2
    com.ibm.ws.microprofile.openapi.fat.annotations.AnnotationProcessingTest
    1 0 1 0s 1 (*1.4)

io.openliberty.jcache.internal_fat.infinispan FAT tests
10 0 1
  junit.framework.TestSuite
  1 0 1 0s 8
    io.openliberty.jcache.internal.fat.FATSuite
    1 0 1 0s (*2)

(*1.1)
java.lang.RuntimeException: 2022-08-21-06:56:38:404 Timed out waiting for testPrivateFieldVisibility to be in state STARTED. Actual state: STOPPED, Last log message:null
	at componenttest.topology.impl.LibertyServer.waitForAppState(LibertyServer.java:6027)
	at componenttest.topology.impl.LibertyServer.addInstalledAppForValidation(LibertyServer.java:6039)
	at com.ibm.websphere.simplicity.ShrinkHelper.exportDropinAppToServer(ShrinkHelper.java:197)
	at com.ibm.ws.microprofile.openapi.fat.annotations.AnnotationProcessingTest.testPrivateFieldVisibility(AnnotationProcessingTest.java:82)

(*1.2)
junit.framework.AssertionFailedError: 2022-08-21-06:56:41:514 Expected: schema named NonPublicFieldVisibilityDataObject to exist
but openApi document was: { ... }
	at com.ibm.ws.microprofile.openapi.fat.annotations.AnnotationProcessingTest.assertDataMatchesSchema(AnnotationProcessingTest.java:165)
	at com.ibm.ws.microprofile.openapi.fat.annotations.AnnotationProcessingTest.testNonPublicFieldVisibility(AnnotationProcessingTest.java:110)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

(*1.3)
java.lang.RuntimeException: 2022-08-21-07:02:43:781 Timed out waiting for testRequiredCustomNameField to be in state STARTED. Actual state: STOPPED, Last log message:null
	at componenttest.topology.impl.LibertyServer.waitForAppState(LibertyServer.java:6027)
	at componenttest.topology.impl.LibertyServer.addInstalledAppForValidation(LibertyServer.java:6039)
	at com.ibm.websphere.simplicity.ShrinkHelper.exportDropinAppToServer(ShrinkHelper.java:197)
	at com.ibm.ws.microprofile.openapi.fat.annotations.AnnotationProcessingTest.testRequiredCustomNameField(AnnotationProcessingTest.java:123)

(*1.4)
java.lang.Exception: Errors/warnings were found in server AnnotationProcessingServer logs:
 <br>[8/21/22, 6:50:41:043 PDT] 00000029 com.ibm.ws.app.manager.AppMessageHelper                      E CWWKZ0002E: An exception occurred while starting the application testPrivateFieldVisibility. The exception message was: com.ibm.ws.container.service.state.StateChangeException: org.jboss.weld.exceptions.DefinitionException
 <br>[8/21/22, 6:56:42:274 PDT] 0000002b com.ibm.ws.app.manager.AppMessageHelper                      E CWWKZ0002E: An exception occurred while starting the application testRequiredCustomNameField. The exception message was: com.ibm.ws.container.service.state.StateChangeException: org.jboss.weld.exceptions.DefinitionException
	at componenttest.topology.impl.LibertyServer.checkLogsForErrorsAndWarnings(LibertyServer.java:3003)

(*2)
java.lang.IllegalStateException: Used testcontainer image(s) [infinispan/server:13.0.10.Final] were not found in the fat.test.container.images property! To correct this, add fat.test.container.images: infinispan/server:13.0.10.Final to the bnd.bnd file for this FAT so that a testcontainer image graph can be generated in the future.
	at componenttest.containers.ImageVerifier.assertImages(ImageVerifier.java:66)
	at componenttest.containers.TestContainerSuite.teardownTestContainers(TestContainerSuite.java:53)
```

Rerunning personal build.

**Comment by tbitonti** [comment]
#build

**Comment by tbitonti** [comment]
Failure 1.1 matches RTC defect 292234.

**Comment by tbitonti** [comment]
Failure 2 is similar but not exactly the same as RTC defect 291450.

**Comment by tbitonti** [comment]
#build

File: dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/ApplicationConfigurator.java [hunk 1]
```diff
@@ -868,6 +859,11 @@ public void updated(String pid, Dictionary<String, ?> properties) throws Configu
         try {
             synchronized (this) {
                 ApplicationConfig appConfig = new ApplicationConfig(pid, properties, _applicationManager);
+
+                (new Throwable("Update to PID [ " + pid + " ]" +
+                              " ID [ " + appConfig.getConfigProperty("id") + " ]" +
+                              " Name [ " + appConfig.getName() + " ]")).printStackTrace(System.out);
+
```
Comment (jhanders34) [inline dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/ApplicationConfigurator.java]:
```suggestion            
```

File: dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/statemachine/CacheUtils.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+     * all cases ensures that unix and windows systems have the same cache
+     * IDs.
+     *
+     * @param pid The PID of the application.
+     * @param id The ID property of the application.
+     *
+     * @return The encoded cache ID of the application.
+     */
+    public static String getCacheId(String pid, String id) {
+        return urlEncode((pid == null) ? id : pid);
```
Comment (jhanders34) [inline dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/statemachine/CacheUtils.java]:
```suggestion
        return urlEncode((id == null) ? pid : id);
```

File: dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/statemachine/ApplicationStateMachineImpl.java [hunk 3]
```diff
@@ -1321,13 +1329,14 @@ private void enterState(InternalState state) {
                         synchronized (_stateLock) {
                             resolveFileCallback = new ResolveFileCallback();
                             ApplicationConfig appConfig = _appConfig.get();
+                            String configId = (String) appConfig.getConfigProperty("id");
                             String configPid = appConfig.getConfigPid();
                             String location = appConfig.getLocation();
                             if (isLocationAURL(location)) {
-                                resolveFileAction = new DownloadFileAction(_locAdmin, configPid, location, resolveFileCallback, _handler);
+                                resolveFileAction = new DownloadFileAction(_locAdmin, configId, configPid, location, resolveFileCallback, _handler);
```
Comment (jhanders34) [inline dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/statemachine/ApplicationStateMachineImpl.java]:
```suggestion
                                resolveFileAction = new DownloadFileAction(_locAdmin, configPid, configId, location, resolveFileCallback, _handler);
```

File: dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/statemachine/ApplicationStateMachineImpl.java [hunk 4]
```diff
@@ -1321,13 +1329,14 @@ private void enterState(InternalState state) {
...
                             } else {
                                 ApplicationMonitorConfig appMonitorConfig = _appMonitor.getConfig();
-                                resolveFileAction = new ResolveFileAction(_ctx, appMonitorConfig.getPollingRate(), appMonitorConfig.getUpdateTrigger(), _locAdmin, appConfig.getName(), configPid, location, resolveFileCallback, _handler);
+                                resolveFileAction = new ResolveFileAction(_ctx, appMonitorConfig.getPollingRate(), appMonitorConfig.getUpdateTrigger(), _locAdmin, appConfig.getName(), configId, configPid, location, resolveFileCallback, _handler);
```
Comment (jhanders34) [inline dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/statemachine/ApplicationStateMachineImpl.java]:
```suggestion
                                resolveFileAction = new ResolveFileAction(_ctx, appMonitorConfig.getPollingRate(), appMonitorConfig.getUpdateTrigger(), _locAdmin, appConfig.getName(), configPid, configId, location, resolveFileCallback, _handler);
```

File: dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/DeploymentTest.java [hunk 5]
```diff
@@ -72,10 +73,10 @@ public class DeploymentTest {
     public static LibertyServer server;
 
     @ClassRule
-    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME,
-                                                             MicroProfileActions.MP60, // mpOpenAPI-3.1, LITE
-                                                             MicroProfileActions.MP50, // mpOpenAPI-3.0, FULL
-                                                             MicroProfileActions.MP41);// mpOpenAPI-2.0, FULL
+    public static RepeatTests r = OpenApiActions.repeat(SERVER_NAME,
+                                                        OpenApiActions.MP_OPENAPI_31, // mpOpenAPI-3.1, LITE
```
Comment (jhanders34) [inline dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/DeploymentTest.java]: This merge didn't go well.  It should be the previous code now.

**Review by jhanders34 — CHANGES_REQUESTED** [review]

**Comment by tbitonti** [comment]
FAT Tests Added:

- test.server.config.CacheIdRestartTest.testCacheIds_Restart()
  - Verify default cache ID assignment.  Verify use of explicit application IDs as cache IDs.
- test.server.config.CacheIdUpdateTest.testCacheIds_Update()
  - Update application configuration to add explicit IDs.  Verify the change of IDs from default IDs to explicit IDs.
- test.server.config.CacheIdRearrangeTest.testCacheIds_Rearrange()
  - Update application configuration to change the order of application configurations.  Verify the correct application function despite the exchange of default ID assignments across server restarts.

**Comment by tbitonti** [comment]
#build

**Comment by tbitonti** [comment]
#libby

File: dev/com.ibm.ws.config_fat/test-applications/cacheid1/resources/META-INF/permissions.xml [hunk 6]
```diff
@@ -0,0 +1,80 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2017 IBM Corporation and others.
```
Comment (jhanders34) [inline dev/com.ibm.ws.config_fat/test-applications/cacheid1/resources/META-INF/permissions.xml]:
```suggestion
    Copyright (c) 2022 IBM Corporation and others.
```

File: dev/com.ibm.ws.config_fat/test-applications/cacheid0/resources/META-INF/permissions.xml [hunk 7]
```diff
@@ -0,0 +1,80 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2017 IBM Corporation and others.
```
Comment (jhanders34) [inline dev/com.ibm.ws.config_fat/test-applications/cacheid0/resources/META-INF/permissions.xml]:
```suggestion
    Copyright (c) 2022 IBM Corporation and others.
```

File: dev/com.ibm.ws.config_fat/publish/servers/com.ibm.ws.config.cacheid.update/bootstrap.properties [hunk 8]
```diff
@@ -0,0 +1,12 @@
+###############################################################################
+# Copyright (c) 2017 IBM Corporation and others.
```
Comment (jhanders34) [inline dev/com.ibm.ws.config_fat/publish/servers/com.ibm.ws.config.cacheid.update/bootstrap.properties]:
```suggestion
# Copyright (c) 2022 IBM Corporation and others.
```

File: dev/com.ibm.ws.config_fat/publish/servers/com.ibm.ws.config.cacheid.restart/bootstrap.properties [hunk 9]
```diff
@@ -0,0 +1,12 @@
+###############################################################################
+# Copyright (c) 2017 IBM Corporation and others.
```
Comment (jhanders34) [inline dev/com.ibm.ws.config_fat/publish/servers/com.ibm.ws.config.cacheid.restart/bootstrap.properties]:
```suggestion
# Copyright (c) 2022 IBM Corporation and others.
```

File: dev/com.ibm.ws.config_fat/publish/servers/com.ibm.ws.config.cacheid.rearrange/bootstrap.properties [hunk 10]
```diff
@@ -0,0 +1,12 @@
+###############################################################################
+# Copyright (c) 2017 IBM Corporation and others.
```
Comment (jhanders34) [inline dev/com.ibm.ws.config_fat/publish/servers/com.ibm.ws.config.cacheid.rearrange/bootstrap.properties]:
```suggestion
# Copyright (c) 2022 IBM Corporation and others.
```

File: dev/com.ibm.ws.config_fat/fat/src/test/server/config/CacheIdUpdateTest.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+//    verifyCacheIds(server, testName, "revert-0", ids_update1);
+//
+//    server.setMarkToEndOfLog();
+//    FileRewriter.update(configPath, WAR1_ASSIGNED_ID, WAR1_DEFAULT_ID);
+//    Thread.sleep(200);
+//    waitForApp(APP_NAME1);
+//
+//    verifyApp0(server);
+//    verifyApp1(server);
+//    verifyCacheIds(server, testName, "revert-1", ids_update1);
```
Comment (jhanders34) [inline dev/com.ibm.ws.config_fat/fat/src/test/server/config/CacheIdUpdateTest.java]: Is this commented out code needed?
```suggestion
```

**Review by jhanders34 — APPROVED** [review]
I am approving it since the review comments I added are not required for the code, but should be made regardless.

**Comment by tbitonti** [comment]
One build failure, which appears to be an infrastructure problem:
```
io.openliberty.microprofile.telemetry.1.0.internal.jaeger_fat FAT tests | 2 | 0 | 1 |   |  
junit.framework.TestSuite | 1 | 0 | 1 | 0 s | 2
io.openliberty.microprofile.telemetry.internal.tests.AutoInstrumentationTest

```

**Comment by tbitonti** [comment]
#build

**Comment by tbitonti** [comment]
#libby

**Comment by tbitonti** [comment]
#build

**Comment by tbitonti** [comment]
Successful field test.  Merging.

**Pull request merged**